Fix Part of #4938: Redesign Profile login and PIN Reset(1/12)

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fixes Part of #4938 
This PR introduces the following changes:

- redesigns the profile login screen 
- ensures that profiles with pin will login via this new screen when the onboardingv2 feature flag is enabled
- handles PIN reset when the user has forgotten their PIN
- expands support for Jetpack Compose
- adds tests for all these changes

## Detailed Explanation of the Changes
### Use of Jetpack Compose
We agreed that we can use Jetpack Compose for new screens being introduced, complete with its own activity and fragment. Although there is an existing login screen, the UI and flows are a little different from the new one, so I opted to create an new screen. The legacy screen has TODOs added to earmark it for removal.

The PR introduces Compose UIs for the screen, as well as the admin pin reset dialogs, since these are alert dialogs. The non-admin pin reset dialog reuses the existing dialogs, since they are standalone dialog fragments.

The Composable functions are private because they are not being tested directly, like you would in in a purely Compose project. The two admin PIN reset dialogs are however in their own files because it made sense to isolate them at that level, being that they are independent components.

Finally, I added the `"androidx.compose.runtime:runtime-livedata": "1.1.1",` dependency to be able to observe livedata objects from within composables.

### Test Design
Because the screen contains both Compose and non-compose components, some of the tests use both Compose nodes and Espresso Matchers and ViewInteractions. I am pleased with how well the dialogs tested.

I used `createEmptyComposeRule()` as the test rule because I found out previously in https://github.com/oppia/oppia-android/pull/5387 that it is the most appropriate rule to use when there is need to launch a new activity for every test, which in our case is useful for testing with different feature flag states.

### `maven_install.json`
Due to conflicts with develop, I had to regenerate this file since we can't directly edit it.

### Misc
There are no dark mode screenshots since the design team is still working on those(https://github.com/oppia/design-team/issues/192), and we will need to include them in a cleanup PR.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
||Admin| Non-Admin|
|---|---|---|
|Screenshot|![Screenshot_2025-05-23-21-59-15-096_org oppia android](https://github.com/user-attachments/assets/217c5054-0429-4e2f-92b1-fdbe98966fe3)|![Screenshot_2025-05-23-21-59-20-605_org oppia android](https://github.com/user-attachments/assets/65f32b07-620b-4081-971c-3ff0bf289daa)|
|Happy Path|https://github.com/user-attachments/assets/e5201d13-cd8f-476f-b3f8-35a830aa0074|https://github.com/user-attachments/assets/53ee4e37-f889-4bd6-9b08-0d3ef60b6c44|
|Wrong PIN|https://github.com/user-attachments/assets/91613b98-4b67-4f89-87dd-e32a306e96e8|https://github.com/user-attachments/assets/4d705cd9-4821-443b-9ce9-47d1a8a61989|
|Reset PIN|https://github.com/user-attachments/assets/eee181f0-4aa7-4b2d-88fa-8bd42fc39a69|https://github.com/user-attachments/assets/e061c064-8382-4afe-99ba-84e4101cf0ed|

adhiamboperes committed:
New layout for auth

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into redesign-pin-input

# Conflicts:
#	app/src/main/res/values/styles.xml

adhiamboperes committed:
Add compose livedata dependency

adhiamboperes committed:
Create color and string resources

adhiamboperes committed:
Create xml layouts

adhiamboperes committed:
Create ProfileLoginActivity

adhiamboperes committed:
Create ProfileLoginFragment

adhiamboperes committed:
Refactor login intent to new ProfileLoginActivity

adhiamboperes committed:
Add todos to track cleanup

adhiamboperes committed:
Cleanup unused items

adhiamboperes committed:
Add admin reset pin flow

adhiamboperes committed:
Add route to non-admin reset pin flow

adhiamboperes committed:
Add non-admin pin reset flow

adhiamboperes committed:
wip commit

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Add activity tests

adhiamboperes committed:
Add fragment tests

adhiamboperes committed:
Break down large composables

adhiamboperes committed:
Add tests for login flow

adhiamboperes committed:
Move dialogs to own files

adhiamboperes committed:
Add tests for dialog flows: wip for non-admin

adhiamboperes committed:
Add todo to ensure an admin profile is created when all profile data is reset

adhiamboperes committed:
Fix lint issues

adhiamboperes committed:
Fix static analysis checks failures

adhiamboperes committed:
Fix ProfileChooserFragmentTest failures

adhiamboperes committed:
Finish dialog tests

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Format dagger modules

adhiamboperes committed:
Fix missing dagger module

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ConfirmDataResetDialog.kt:
```diff
@@ -0,0 +1,63 @@
+      dismissOnClickOutside = false,
+      dismissOnBackPress = false
+    ),
+    onDismissRequest = {
```
You can simplify this with a function reference, its cleaner and prevents unnecessary lambda call: (Optional)

`onDismissRequest = onDismissRequest`

adhiamboperes replied:
Ditto here.

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ConfirmDataResetDialog.kt:
```diff
@@ -0,0 +1,63 @@
+    },
+    dismissButton = {
+      TextButton(
+        onClick = {
```
Same as above — consider refactoring : (Optional)
TextButton(onClick = onDismissRequest)

adhiamboperes replied:
Ditto here.

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ConfirmDataResetDialog.kt:
```diff
@@ -0,0 +1,63 @@
+      }
+    },
+    confirmButton = {
+      TextButton(
```
Same as above — consider refactoring : (Optional)
TextButton(onClick = deleteAppData)

adhiamboperes replied:
Same comment regarding the Unit.

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ForgotAdminPinDialog.kt:
```diff
@@ -0,0 +1,58 @@
+      dismissOnClickOutside = false,
+      dismissOnBackPress = false
+    ),
+    onDismissRequest = {
```
Same as above — consider refactoring : (Optional)
`onDismissRequest = onDismissRequest`

adhiamboperes replied:
Ditto here

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ForgotAdminPinDialog.kt:
```diff
@@ -0,0 +1,58 @@
+      onDismissRequest()
+    },
+    dismissButton = {
+      TextButton(
```
Same as above — consider refactoring : (Optional)
TextButton(onClick = onDismissRequest)

adhiamboperes replied:
Ditto here

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/ForgotAdminPinDialog.kt:
```diff
@@ -0,0 +1,58 @@
+      }
+    },
+    confirmButton = {
+      TextButton(
```
Same as above — consider refactoring : (Optional)
TextButton(onClick = onConfirmation)

adhiamboperes replied:
I am unable to remove the lambdas because the `onClick` function expects a unit, which needs the lambda. What I have done, however, is to refactor the lambda to a single line to improve readability.

![Screenshot 2025-07-01 at 13 12 19](https://github.com/user-attachments/assets/93237414-fbb2-48d2-8347-294c8b876ce0)

deonwaju commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -147,7 +147,7 @@ class PinPasswordActivityPresenter @Inject constructor(
           activity.supportFragmentManager.beginTransaction().remove(previousFrag).commitNow()
         }
         val dialogFragment = AdminSettingsDialogFragment
```
Consider using null checks instead of Using !! (not-null assertions),it is risky because if `adminPin` or `name.get()` is null, the app will crash at runtime. 

```
val name = pinViewModel.name.get()
if (adminPin != null && pinViewModel.name.get() != null) {
  val dialogFragment = AdminSettingsDialogFragment.newInstance(adminPin, profileId, pinViewModel.name.get())
  dialogFragment.show(activity.supportFragmentManager, TAG_ADMIN_SETTINGS_DIALOG)
}
```

adhiamboperes replied:
Done

adhiamboperes committed:
fix trim spannable bug

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into develop

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt
#	app/src/main/res/values-night/color_palette.xml
#	app/src/main/res/values/color_palette.xml
#	app/src/main/res/values/strings.xml
#	third_party/maven_install.json

adhiamboperes committed:
Generate a new maven_install due to conflicts in the old one.

adhiamboperes committed:
regenerate maven_install.json

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
This change should not be part of this PR.

adhiamboperes replied:
Fixed

adhiamboperes committed:
address reviewer comment.

adhiamboperes committed:
Revert unrelated changed

adhiamboperes committed:
fix tests

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	app/src/main/res/values/color_palette.xml
#	third_party/maven_install.json

adhiamboperes commented:
Thanks @deonwaju! I have addressed your comments. Please resolve if you have the permissions, otherwise let me know and I will resolve them myself.

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Repin dependencies

adhiamboperes committed:
Merge branch 'redesign-pin-input' of github.com:oppia/oppia-android into redesign-pin-input

BenHenning commented on app/src/main/res/layout/profile_login_activity.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Are we ever adding anything to this layout dynamically? If not, can we just make the `FrameLayout` the root, instead? It helps reducing the layout nesting.

adhiamboperes replied:
Not really. I have flattened the layout.

BenHenning commented on app/src/main/res/layout/profile_login_fragment.xml:
```diff
@@ -0,0 +1,15 @@
+        android:background="@color/component_color_shared_screen_primary_background_color"
+        android:gravity="center">
+
+        <androidx.compose.ui.platform.ComposeView
```
Similar to my comment for the `profile_login_activity` layout, can the `ComposeView` possibly be the top-level one here? I'm assuming the `gravity` doesn't matter since the `ComposeView` has a `match_parent` width/height, and we should be able to set the background of the `ComposeView` (I'm assuming).

adhiamboperes replied:
Changed the top-level view to be the ComposeView

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ConfirmDataResetDialog.kt:
For this & the other class: we need to start thinking about naming standards for these files since they're a new type of file altogether that we haven't really had before.

https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md#naming-unit-composable-functions-as-entities explains this, but it's honestly quite a bit more complicated than I was expecting. I suspect we'll need to think about this and have a lot more examples before we can settle on naming conventions that make sense to us.

I don't think there's an action item here, but I did want to get your thoughts on the matter, if you have any.

adhiamboperes replied:
Generally, I believe that naming composable entities should follow the same conventions as the existing files in the project. We typically have FeatureNameActivity/Fragment/Dialog, and for custom views, we have CustomView. Composable entities will likely fall into 3 categories:
1. the composable functions that represent small view units, and will be written inside a presenter(or a larger Composable class, e.g. https://github.com/oppia/oppia-android/blob/43a08262fa9d5abddb9f0c5848aacbe6ca9af680/app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt#L207
2. composables that represent screens, standalone componets such as dialogs
3. and finally, custom views, e.g e.g. [AllClassroomsHeaderText.kt](https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt)

The list is not exhaustive, but will form a bulk of Composables that we write.

Combining our present naming convention with the Noun imperative from the docs, it shouldn't be too complex.

- `ComingSoonTopicList.kt` would be renamed to `ComingSoonTopicListView.kt`
- This dialog can be renamed from `ConfirmDataResetDialog.kt` to `DataResetConfirmationDialog`, 
- and the smaller unit composables can be named explicitly after the view they represent, e.g `NameErrorText()`

BenHenning replied:
That makes sense, though my concerns right now:
- Traditionally in Android naming, ending something with 'Activity', 'Fragment', 'Dialog', 'View', etc. usually meant that it was a subclass of the corresponding thing. With this naming style, that's no longer true and could introduce some confusion, especially in the intermediary state before we move entirely over to Compose. Do other apps just not care about this difference?
- The smaller unit composables still seem very fuzzy to me, but if it's possible for us to limit what they can be called to select types of composables (e.g. 'Text' in your example, and others) then we could perhaps tighten the naming quite a bit. This is a case where I think we would be advantaged with a custom CI check to ensure naming consistency. I can see things quickly getting very confusing without such a check in place.

adhiamboperes replied:
Most apps I’ve seen that have moved to Compose don’t stress too much about the inheritance mismatch. They often:

- Use `*Screen`, `*Dialog`, `*Card`, `*Row`, etc. for composables.
- Rely on package structure (e.g. ui/screen, ui/component, ui/dialog) to disambiguate.
- Accept that the suffix signals role, not class hierarchy anymore.

**Possible conventions for Oppia**
Screens (top-level composables) - represents a whole navigable screen.
Naming: `XyzScreen.kt`,  e.g `ProfileListScreen()`.

b. Dialogs & other standalone components - represents:a self-contained UI element invoked directly (dialogs, bottom sheets).
Naming: `XyzDialog.kt`,  e.g. `DataResetConfirmationDialog()`.
This keeps alignment with our current convention while signaling functionality not inheritance.

c. Small reusable UI units (building blocks)
Naming: `XyzView.kt` or `XyzSomething()` where Something is a UI role.
e.g `ErrorText()`, `TopicCard()`.

This way, Dialog and Screen still mean “top-level entry point,” while View and specific UI-role suffixes capture small building blocks.


I agree CI checks will be invaluable here. We can enforces that only approved suffixes (Screen, Dialog, View, Text, Card, etc.) allowed for composables, and that File name ↔ top-level composable name match.

To avoid confusion in the hybrid state, we can separate Compose files into dedicated packages (e.g. ui/compose/dialog vs ui/dialog).

BenHenning replied:
I like where this discussion is headed--perhaps it's a good time to maybe move this into a document for further collaboration?

I don't want to over-index on what other apps are doing unless there's clear, unambiguous best practices that we can follow. It doesn't sound like there is and that people are largely following an unspecified or partly specified convention. That convention demonstrably has gaps that could lead to confusion, and I think your suggestions largely address those.

I also think there may be a bit more we can do on things that call into (c) to fully define them (maybe having a suffix like 'Unit' or something, not sure), but this is best continued out of this PR, I think, so that it doesn't block the PR itself.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginActivityPresenter.kt:
```diff
@@ -0,0 +1,81 @@
+
+  /** Handles showing the [ResetPinDialogFragment]. */
+  fun handleRouteToResetPinDialog(profileId: ProfileId, profileName: String) {
+
```
Nit: please remove empty line for convention.

adhiamboperes replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginActivityPresenter.kt:
```diff
@@ -0,0 +1,81 @@
+      as DialogFragment
+    resetPinDialog.dismiss()
+
+    fragmentManager.executePendingTransactions()
```
For here & above: why is this needed?

adhiamboperes replied:
Since there is a series of dialogs leading to one another, forcing all pending transactions to complete immediately helps prevent stacking and errors.

BenHenning replied:
Got it--suggest adding a line comment explaining the need for the function.

Do tests cover it missing? (It may not be easy to test--that's fine, it just means that the line comment is more important to guard against unintentional regressions in the future).

BenHenning replied:
Reminder to please respond to this.

adhiamboperes replied:
Apologies -- I missed this. I have added a comment.

The tests as they are, pass with or without this function, but I am not sure how to test the posibility of stacking dialogs currently. In future, if we do end up with a pure Compose app, we might be able to test such scenarios.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! I haven't looked at the bulky bits of the PR yet (the new `Composable`s, the new fragment presenter, and the fragment test), but did look at everything else and left some comments. Besides those, could you also fix CI & look at the replies below?

> ### Other Design Choices
> **Dialog State Management:**
> 
> * When opening two dialogs in sequence in compose, I chose to display the second dialog without dismissing the first one. This causes the second dialog to appear on top of the first. When the second dialog is dismissed, the first remains open and must be manually closed.
> * On the flip side, if the first dialog is configured to close when the positive button is clicked(to open the next dialog), it dismisses correctly, but then the second dialog doesn’t appear at all.
> 
> Neither approach is ideal, but keeping the first dialog open while showing the second is the more acceptable of the two.

This shouldn't be necessary, I think. In the non-Compose world you can check if the fragment has already been added to the activity's fragment manager (via its tag) and replace it with a new one (or not open a new one at all) to de-duplicate the dialogs. Shouldn't there be a way to do the same via Compose?

> ### Misc
> There are no dark mode screenshots since the design team is still working on those([oppia/design-team#192](https://github.com/oppia/design-team/issues/192)), and we will need to include them in a cleanup PR.

Is there a tracking issue for this that includes the details for updating this & any other affected flows in particular?

> ## Essential Checklist
> ...

Can you please fill this out?

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Exempt files from coverage.

adhiamboperes commented:
> > ### Other Design Choices
> > **Dialog State Management:**
> 
> This shouldn't be necessary, I think. In the non-Compose world you can check if the fragment has already been added to the activity's fragment manager (via its tag) and replace it with a new one (or not open a new one at all) to de-duplicate the dialogs. Shouldn't there be a way to do the same via Compose?

Compose Navigation replaces Fragment transactions and tags entirely, but has state. For hiding/showing Composables, we use `remember { mutableStateOf(false) }`, and set the value to true when we want the Composable to be added to the tree, and false when we want it to be removed. I did some refactor, and found a way to manage these dialogs correctly.



> > ### Misc
> > There are no dark mode screenshots since the design team is still working on those([oppia/design-team#192](https://github.com/oppia/design-team/issues/192)), and we will need to include them in a cleanup PR.
> 
> Is there a tracking issue for this that includes the details for updating this & any other affected flows in particular?

#4938 should track all the work related to this project.



> > ## Essential Checklist
> > ...
> 
> Can you please fill this out?


Done

adhiamboperes committed:
Address reviewer comments

adhiamboperes committed:
Fix coverage failures

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/AdminPinRecoveryDialog.kt:
```diff
@@ -0,0 +1,52 @@
+import org.oppia.android.app.ui.R
+
+/** Test tag for the admin's forgot pin dialog. */
+const val ADMIN_PIN_RECOVERY_DIALOG_TEST_TAG = "TEST_TAG.admin_forgot_pin"
```
This seems a bit gross to me--it's generally a bad practice to customize something that affects the real production behavior of a component for test-specific reasons. Is this truly the cleanest way to do this?

We never ran into this problem pre-Compose since most fragments would have tags that had a legitimate use case for transactions.

Another thought: could we inspect that _a_ dialog is open, and that it has the expected text, rather than validating against the specific tag? This is more akin to what an end-to-end test (that is, a user manually verifying it) would do.

Additional thought after reviewing the tests: I definitely am not keen on this approach. It places far too much coupling between the test logic and the UI arrangement. As far as I can tell, it's probably possible to show _something_ with the tag that isn't at all the thing we actually care about. That's why we largely validate using strings in most places, but sometimes IDs (which is admittedly not much better than a test-only tag, especially if we only add the ID for the purpose of testing). Perhaps this is the "new way" of doing things and we need to rely more heavily on future screen shot tests for correct layout and text display...but I'd rather we explore other options before hard committing to that outcome.

adhiamboperes replied:
`Composables` don't have IDs and you can't use the Node numbers shown in the semantics tree to match them. If matching a node with its semantics properties is impractical or impossible, such as in the case of the PIN boxes which are dynamically generated and do not contain text when created, we can use the [testTag](https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/package-summary#(androidx.compose.ui.Modifier).testTag(kotlin.String)) modifier as a last resort. For this reason, I have removed the test tags where there is an alternate, more traditional matcher available.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/DataResetConfirmationDialog.kt:
```diff
@@ -0,0 +1,59 @@
+import org.oppia.android.app.ui.R
+
+/** Test tag for the confirm reset app data dialog. */
+const val DATA_RESET_CONFIRMATION_DIALOG_TEST_TAG = "TEST_TAG.admin_reset_app_data"
```
Ditto for this test tag.

adhiamboperes replied:
Removed.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginFragmentPresenter.kt:
```diff
@@ -0,0 +1,447 @@
+      .observe(fragment) {
+        if (it is AsyncResult.Success) {
+          activity.startActivity(
+            if (enableMultipleClassrooms.value)
```
Nit: `if`-`else` here needs braces since it's wrapping to multiple lines.

adhiamboperes replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginFragmentPresenter.kt:
```diff
@@ -0,0 +1,447 @@
+        fontSize = 16.sp,
+        color = colorResource(id = R.color.component_color_profile_login_primary_text_color)
+      ),
+      modifier = Modifier.testTag(PROMPT_TEST_TAG)
```
Ditto for this (any any other) test tag (see my earlier comment in the PR).

adhiamboperes replied:
Removed all test tags except for `PIN_INPUT_TEST_TAG` and `PIN_BOX_TEST_TAG` as explained above.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginFragmentPresenter.kt:
```diff
@@ -0,0 +1,447 @@
+        .aspectRatio(0.7F)
+        .border(
+          width = if (isFocused) 2.dp else 1.dp,
+          color = if (isFocused)
```
Ditto here (`if`-`else` needs braces). Ditto elsewhere in the PR (where needed).

adhiamboperes replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileLoginFragmentPresenter.kt:
```diff
@@ -0,0 +1,447 @@
+        setShakeOffset(0f)
+
+        // Clear the incorrect PIN and hide the error message after a 2s delay.
+        delay(2000)
```
Does this & the above block the UI thread, or is Compose properly handing back control? Otherwise we risk an ANR (which is why we use the lifecycle-safe timer utility for such cases).

adhiamboperes replied:
No. Please see [documentation](https://developer.android.com/develop/ui/compose/side-effects#launchedeffect).

- `LaunchedEffect` launches a coroutine in the composition’s `CoroutineScope`, tied to the current `CompositionLifecycle`.
- The coroutine runs on the main thread (Dispatchers.Main.immediate) by default.
- Every delay(...) is non-blocking, it suspends the coroutine and returns control to the main thread’s event loop.
- The coroutine will be cancelled if LaunchedEffect leaves the composition.

BenHenning replied:
Wow. The cooperation between the main thread's coroutine scope and the main looper must be very complex to make something like that work. I...wonder if anyone's run into challenging threading issues from that mechanism. At a glance this seems very cool since it solves a rather annoying problem with Android development, but I do worry a bit what the actual costs are to make this work.

Either way, I don't have any suggestions to change this since it's obviously following best practices and the cooperative dispatcher _should_ avoid ANRs.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+      .assertIsDisplayed()
+    composeRule.onNodeWithTag(PROMPT_TEST_TAG).assertIsDisplayed()
+      .assertTextContains(context.getString(R.string.profile_login_activity_enter_pin_prompt))
+      .assertIsDisplayed()
```
Checking that it's displayed twice?

adhiamboperes replied:
Yes. Fixed here and elsewhere.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+  fun testFragment_onLaunch_allTextViewsHaveCorrectContent() {
+    setUpTestApplicationComponentWithFeatureFlags()
+    profileTestHelper.addOnlyAdminProfile()
+    scenario = launch(ProfileLoginActivity::class.java)
```
This seems...ill-advised. We probably should be closing the scenario--is three a reason you don't approach it that way here (like we do in most/all other UI tests)?

adhiamboperes replied:
I wrapped the tests in the `use {}` block like we do in other tests. Unfortunately, this has resultled into a lot of formatting changes in the file due to new indentation.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+
+    repeat(5) { index ->
+      composeRule
+        .onNodeWithTag(PIN_BOX_TEST_TAG + index, useUnmergedTree = true)
```
What does `useUnmergedTree` do? The documentation made it sound like it had more to do with the error output during failure, but I admittedly only looked into it a ltitle.

adhiamboperes replied:
compose tests use the [semantics tree](https://developer.android.com/jetpack/compose/testing#semantics) to look for elements on the screen and read their properties. This is the structure that accessibility services use as well, as they're meant to be read by a service such as [TalkBack](https://support.google.com/accessibility/android/answer/6283677).

In newer versions of Compose, you can use `printToLog` which will output the full semantics tree showing all the nodes in that composable. This is similar to the entire view hierachy output printed by Robolectric  when a test fails with `cannot find view in hierachy matching...`. e.g.

```php
 |-Node #3 at (l=42.0, t=105.0, r=105.0, b=168.0)px
        | Role = 'Tab'
        | Selected = 'false'
        | StateDescription = 'Not selected'
        | ContentDescription = 'Overview'
        | Actions = [OnClick]
        | MergeDescendants = 'true'
        | ClearAndSetSemantics = 'true'
```

The Semantics tree always tries to be as compact as possible, showing only the information that is relevant. The above node has property `MergeDescendants = 'true'` which means that this node had descendants, but they have been merged into it, but in some tests, we want to access all the nodes.  Passing` useUnmergedTree = true`  to the onRoot finder allows us to query the unmerged Semantics tree. All finders have the `useUnmergedTree` parameter.

Sample: https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/samples/ui/test/samples/SemanticsNodeInteractionSamples.kt;drc=03ca30d22e6ee3483142f2e4048db459cb5afb79;l=58

BenHenning replied:
Interesting. I like that--it adds a lot more flexibility in testing.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+  }
+
+  @Test
+  fun testFragment_nonAdmin_enterTwoDigits_doesNotTriggerLogin() {
```
Should this also be verifying no intent is fired? That would actually be verifying that the login hasn't been triggered.

(I was expecting more consistency with the next test, including the `runCurrent` call after inputting text).

Ditto for the admin case.

adhiamboperes replied:
Done for asserting that no intent is fired in the 2 test, but `runCurrent` does not need to be called because nothing happens after this input, therefore no synchronization is required.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+    scenario = launch(ProfileLoginActivity::class.java)
+    testCoroutineDispatchers.runCurrent()
+
+    composeRule.onNodeWithTag(FORGOT_PIN_TEST_TAG).performClick()
```
This feels really fake to me, and perhaps leans more heavily into my concerns with using test-only tags.

A user cannot perform this action, can they? We normally would require them to enter the wrong pin, first, _then_ click on the 'forgot pin' button. Here, we're clicking on a button that's not even visible.

adhiamboperes replied:
The behaviour tested here is correct and consistent with the legacy flow. The forgot PIN button is displayed when the screen is rendered and can be clicked without first entering an incorrect PIN.

adhiamboperes replied:
The test cannot click on a component that is not in the semantics tree, even if using the test tag to try to access it. The test will fail.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+      )
+      .performClick()
+
+    composeRule.onNodeWithTag(DATA_RESET_CONFIRMATION_DIALOG_TEST_TAG)
```
Suggest also verifying the app isn't closing (for consistency with the next test).

adhiamboperes replied:
Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+
+    composeRule.onNodeWithTag(FORGOT_PIN_TEST_TAG).performClick()
+
+    composeRule.waitForIdle()
```
Huh. What does this do? Does it actually interact with Espresso `IdlingResource` or is it doing something else?

I foresee me having to add Jetpack Compose to #89 if there's another asynchronous mechanism that needs to settle for tests.

adhiamboperes replied:
This method waits for all `IdlingResources` to become idle before executing actions.

https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-test-junit4/src/jvmMain/kotlin/androidx/compose/ui/test/junit4/ComposeTestRule.jvm.kt;drc=e687da9e7221a7de5adac9d3150334975233821d;l=98

Although similar to [Espresso's Idling Resources](https://developer.android.com/training/testing/espresso/idling-resource), [Compose's Idling Resources](https://developer.android.com/reference/kotlin/androidx/compose/ui/test/IdlingResource) is a different API.

The `waitForIdle()` function here manually synchronizes the test to wait for Compose to become idle in order for the rest of the test (which runs on Espresso) to continue executing.

BenHenning replied:
Thanks.

Updated #89 to consider this.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/ProfileLoginFragmentTest.kt:
```diff
@@ -0,0 +1,829 @@
+  }
+
+  @Test
+  fun testFragment_nonAdminUser_enterAndSubmitNewPin_opensSuccessDialog() {
```
Suggest perhaps adding one more test that completes the flow here and validates that the user can actually log in with the new PIN once reset (that seems like a nice completion of the set of steps being verified here). It is slightly "end-to-end-esque" but I think that's fine here.

adhiamboperes replied:
Done.

BenHenning commented on app/src/main/res/layout/profile_login_fragment.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <androidx.compose.ui.platform.ComposeView
```
Nit: This should be two-spaced, I think.

adhiamboperes replied:
Done

BenHenning reviewed (commented):
Thanks @adhiamboperes! Apologies again for the delayed review. I plan to be a lot more responsive for the rest of the PRs in this project.

I've taken a pass on everything now, and have a bunch of thoughts. Most of them are around how we are actually testing the new Compose layouts--I have a few specific concerns that I think need to be addressed.

Most of my remaining comments are nits or small suggestions.

PTAL.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Use use {} block for scenario

adhiamboperes committed:
Remove test tags

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Merge branch 'redesign-pin-input' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Add lint exemption

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Add comments explaining test tag usage

adhiamboperes committed:
fix kdoc formatting

BenHenning reviewed (commented):
Thanks @adhiamboperes! Just had a few follow-up thoughts, but I like the changes that moved away from test tags where it's avoidable (which is, hopefully, most places as we move more of the app over to Compose).

The PR otherwise largely looks good to me.

BenHenning commented:
Also, it looks like there are some conflicts to address from `develop`.

adhiamboperes committed:
add comment

adhiamboperes committed:
resolve merge conflicts from develop

adhiamboperes committed:
Remove duplicate tests

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into redesign-pin-input

adhiamboperes commented:
@BenHenning, addressed pending comment and resolved conflicts. Also, in the latest commit I have deleted a couple of tests that are actual duplicates of earlier tests. PTAL.

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

BenHenning reviewed (approved):
Thanks @adhiamboperes!
